### ISSUE TITLE
Tectonic is always running 6 times and reaching the maximum number of runs

### ISSUE THREAD

**Issue by mseri** [tectonic-typesetting/tectonic#679]
When compiling my latex projects, tectonic always runs 6 times (`TeX rerun seems needed, but stopping at 6 passes`), justifying each re-run with
```
Rerunning TeX because "aom.run.xml" changed ...
```

I have isolated 3 of those files and computed their hashes, and they all look identical:
```
$ openssl dgst -sha256 < aom.run.xml       
(stdin)= 21f268eb669dce0181d75b69c8fabe9f7b34874df1adfeb1b34ef220cbef35eb
$ openssl dgst -sha256 < aom.run.xml.1 
(stdin)= 21f268eb669dce0181d75b69c8fabe9f7b34874df1adfeb1b34ef220cbef35eb
$ openssl dgst -sha256 < aom.run.xml.2
(stdin)= 21f268eb669dce0181d75b69c8fabe9f7b34874df1adfeb1b34ef220cbef35eb
```

This happens with both 0.3.0 and 0.3.1, I never used earlier versions, so I don't know if it is a regression.
I run `tectonic` on macOS 10.15.7 using the binaries downloaded from the release page.

The document that I used for this report is available at https://github.com/mseri/aom (to avoid font issues I suggest running `sed -i.bak '8,10d' aom.tex`)

**Comment by pkgw** [comment]
Thank you for the detailed report and reproduction information!

It seems that the issue here is that the TeX code that *reads* the XML file only reads a short piece of it (the first 114 bytes or so) when it does so. Tectonic's rerun logic compares the content that was read to the content that was written, so it finds a difference even though the file is not evolving from run to run.

As an experiment, I tried futzing with the logic so that if an input file is closed early, the comparison is performed against the entire contents of the file, as opposed to merely what was read. This seems to fix the issue in this test case.

I'll have to think about it a little bit more, but at first blush I think that the above approach is a sound solution.

**Comment by pkgw** [comment]
Also, the font issues with your build *can* be fixed by specifying your fonts using their filenames in the Tectonic bundle, as opposed to their descriptions, close to:

```
\setmainfont[...]{texgyrepagella-regular.otf}
```

There isn't really any way for people to know what these file names are unless that examine the bundle listing, and you need to use a few special options to deal with bold/italic/etc variants, but it is possible to set up right. I need to document the magic incantation somewhere.

**Comment by pkgw** [comment]
This should be fixed in the upcoming 0.3.3 release, which I will issue shortly.

**Comment by mseri** [comment]
Thank you very much for the fast fix and the suggestion!
